### ISSUE TITLE
Why is model.set_classes(prompts) so slow in YOLOE?

### ISSUE THREAD

**Issue by yang-0201** [ultralytics/ultralytics#24266]
### Search before asking

- [x] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/orgs/ultralytics/discussions) and found no similar questions.

### Question

I encountered an issue when using YOLOE. Although the actual prediction time is quite short, especially in cases with only a few objects, the model.set_classes(prompts) step takes a surprisingly long time. I would really appreciate any clarification on this. Thank you very much.

Here is the situation. I am running inference locally on an RTX 5090.

In the first experiment, I passed all prompts to the YOLOE model at once and performed a single forward pass. The timing results were:

[1/5] setcls=0.6182s | predict=0.7005s | post=0.0196s | encode=0.0030s | total=1.3443s
[2/5] setcls=0.4766s | predict=0.0817s | post=0.0503s | encode=0.0090s | total=0.6221s

From these results, I observed that model.set_classes() alone takes about 0.5 seconds on average, which significantly slows down the overall inference process.

In the second experiment, for performance comparison, I tried iterating over prompts one by one, similar to the SAM3 workflow:

state = processor.set_image(image)

prompt_to_mask = {}

for prompt in prompts:
    processor.reset_all_prompts(state)
    state = processor.set_text_prompt(prompt=prompt, state=state)

In SAM3, the image features can be reused across different prompts. However, because I have multiple prompts for each image, the overhead becomes even more significant in YOLOE. The timing results looked like this:

[4/5] prompts=8 | setcls=4.2097s | predict=0.2499s | post=0.0257s | encode=0.0050s | total=4.5028s

So I would like to ask the Ultralytics team: is this high overhead in set_classes() expected behavior in YOLOE, or is there some way to accelerate it?

**Comment by UltralyticsAssistant** [comment]
👋 Hello @yang-0201, thank you for your interest in Ultralytics 🚀! This is an automated response to help get your issue moving quickly, and an Ultralytics engineer will also assist soon. We recommend a visit to the [Docs](https://docs.ultralytics.com/) for new users where you can find many [Python](https://docs.ultralytics.com/usage/python/) and [CLI](https://docs.ultralytics.com/usage/cli/) usage examples and where many of the most common questions may already be answered.

Since this appears to be a usage/performance ❓ question, please provide as much information as possible to help us investigate, including:
- your `ultralytics` version
- a minimal code snippet showing how `model.set_classes(prompts)` is being called
- whether this timing is from the first run only or also after warmup
- prompt count and prompt format
- any relevant logs or profiler output

If this turns out to be a 🐛 Bug Report, please provide a [minimum reproducible example](https://docs.ultralytics.com/help/minimum-reproducible-example/) to help us debug it.

If this is a custom training ❓ Question, please provide as much information as possible, including dataset image examples and training logs, and verify you are following our [Tips for Best Training Results](https://docs.ultralytics.com/guides/model-training-tips/).

Join the Ultralytics community where it suits you best. For real-time chat, head to [Discord](https://discord.com/invite/ultralytics) 🎧. Prefer in-depth discussions? Check out [Discourse](https://community.ultralytics.com/). Or dive into threads on our [Subreddit](https://www.reddit.com/r/Ultralytics/) to share knowledge with the community.

## Upgrade

Upgrade to the latest `ultralytics` package including all [requirements](https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml) in a [**Python>=3.8**](https://www.python.org/) environment with [**PyTorch>=1.8**](https://pytorch.org/get-started/locally/) to verify your issue is not already resolved in the latest version:

```bash
pip install -U ultralytics
```

## Environments

YOLO may be run in any of the following up-to-date verified environments (with all dependencies including [CUDA](https://developer.nvidia.com/cuda)/[CUDNN](https://developer.nvidia.com/cudnn), [Python](https://www.python.org/) and [PyTorch](https://pytorch.org/) preinstalled):

- **Notebooks** with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo26"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud** Deep Learning VM. See [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/)
- **Amazon** Deep Learning AMI. See [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/)
- **Docker Image**. See [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

## Status

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml/badge.svg" alt="Ultralytics CI"></a>

If this badge is green, all [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule) tests are currently passing. CI tests verify correct operation of all YOLO [Modes](https://docs.ultralytics.com/modes/) and [Tasks](https://docs.ultralytics.com/tasks/) on macOS, Windows, and Ubuntu every 24 hours and on every commit.

**Comment by Y-T-G** [comment]
It's  normal because `set_classes` runs CLIP/MobileCLIP. Why not set all the classes you want to detect at once?

**Comment by yang-0201** [comment]
Thank you, that is a good suggestion. However, my prompts are a large set of open-vocabulary text queries with substantial content, and the test set is unseen, so I am not sure whether setting them all at once would be a reasonable approach.

**Comment by Y-T-G** [comment]
You can probably precalculate the embeddings and then assign it during prediction without recalculating.

**Comment by Y-T-G** [comment]
```python
import torch
from pathlib import Path
from ultralytics import YOLOE

CACHE = Path("text_pe_cache.pt")

def build_cache(model, prompts):
    """Runs CLIP once for the given prompts."""
    tpe = model.get_text_pe(prompts)                   # (1, N, D)
    return dict(zip(prompts, tpe.squeeze(0).cpu()))    # {prompt: (D,)}

def load_or_update(model, prompts):
    cache = torch.load(CACHE, map_location="cpu") if CACHE.exists() else {}
    missing = [p for p in prompts if p not in cache]
    if missing:
        cache.update(build_cache(model, missing))
        torch.save(cache, CACHE)
    return cache

def tpe_from_cache(cache, prompts, device):
    """Rebuild (1, N, D) tensor in the caller's order — no CLIP."""
    return torch.stack([cache[p] for p in prompts]).unsqueeze(0).to(device)


# --- one-time: warm the cache (CLIP runs here) ---
model = YOLOE("yoloe-11s-seg.pt")
cache = load_or_update(model, ["person", "bus", "dog", "bicycle", "traffic light"])

# --- inference: zero CLIP overhead ---
names = ["person", "dog"]
tpe = tpe_from_cache(cache, names, device=next(model.model.parameters()).device)
model.set_classes(names, tpe)
results = model.predict("image.jpg")
```

**Comment by yang-0201** [comment]
> import torch
> from pathlib import Path
> from ultralytics import YOLOE
> 
> CACHE = Path("text_pe_cache.pt")
> 
> def build_cache(model, prompts):
>     """Runs CLIP once for the given prompts."""
>     tpe = model.get_text_pe(prompts)                   # (1, N, D)
>     return dict(zip(prompts, tpe.squeeze(0).cpu()))    # {prompt: (D,)}
> 
> def load_or_update(model, prompts):
>     cache = torch.load(CACHE, map_location="cpu") if CACHE.exists() else {}
>     missing = [p for p in prompts if p not in cache]
>     if missing:
>         cache.update(build_cache(model, missing))
>         torch.save(cache, CACHE)
>     return cache
> 
> def tpe_from_cache(cache, prompts, device):
>     """Rebuild (1, N, D) tensor in the caller's order — no CLIP."""
>     return torch.stack([cache[p] for p in prompts]).unsqueeze(0).to(device)
> 
> 
> # --- one-time: warm the cache (CLIP runs here) ---
> model = YOLOE("yoloe-11s-seg.pt")
> cache = load_or_update(model, ["person", "bus", "dog", "bicycle", "traffic light"])
> 
> # --- inference: zero CLIP overhead ---
> names = ["person", "dog"]
> tpe = tpe_from_cache(cache, names, device=next(model.model.parameters()).device)
> model.set_classes(names, tpe)
> results = model.predict("image.jpg")

Thank you for your suggestion. It successfully solved my problem. After testing, set_classes is no longer a bottleneck, and now almost all of the time is spent on the inference side.

**Comment by glenn-jocher** [comment]
Glad to hear it’s resolved — for repeated open-vocabulary queries, caching the text embeddings once and then calling `model.set_classes(names, embeddings)` is the intended fast path in YOLOE, as shown in the [YOLOE model reference](https://docs.ultralytics.com/reference/models/yolo/model/) and [tasks API](https://docs.ultralytics.com/reference/nn/tasks/).
